### PR TITLE
Fixed #32002 -- Added headers parameter to HttpResponse and subclasses.

### PR DESCRIPTION
**Pull request by knyghty**
[Ticket](https://code.djangoproject.com/ticket/32002)

There are now two overlapping ways to set the `Content-Type` header. I think it makes sense to take `content_type` over `headers`, but not sure if we should check this in `__init__()` and raise an exception, document it, or perhaps neither is necessary.

### CODE REVIEW

File: docs/howto/outputting-csv.txt
Comment (charettes) [inline docs/howto/outputting-csv.txt:25]:
```suggestion
            headers={'Content-Disposition': 'attachment; filename="somefilename.csv"'},
```

File: docs/topics/class-based-views/index.txt
Comment (matthiask) [inline docs/topics/class-based-views/index.txt:120]: This comment should be moved above the `headers=` argument.

File: tests/httpwrappers/tests.py
Comment (matthiask) [inline tests/httpwrappers/tests.py:475]: Maybe test whether `r.headers['x-foo']` is set or something? (`CaseInsensitiveMapping` itself is probably thoroughly tested but it may be useful to verify the case insensitivity here too)

File: docs/topics/class-based-views/index.txt
Comment (knyghty) [reply docs/topics/class-based-views/index.txt]: I think neither place really feels right, but I think it's slightly better there.

File: tests/httpwrappers/tests.py
Comment (knyghty) [reply tests/httpwrappers/tests.py:475]: I'm not exactly sure what it's protecting against, but seems reasonable, have done this.

**Comment by matthiask** [comment]
Looks good! 

I don't want to complicate things more in this pull request BUT have you considered adding support for this argument to `TemplateResponse` and `TemplateSimpleResponse` too? Those two classes use explicit forwarding of arguments, not `*a, **kw`. I think Carlton definitely has a point regarding not adding a headers parameter to the `render()` shortcut, but the template responses are a different case.

**Comment by knyghty** [comment]
I'm not against it but I'll hold off for a second opinion.

Is there some reason those classes don't just pass `*args, **kwargs` through like the other subclasses do?

**Comment by matthiask** [comment]
Not passing `*args` makes sense since it's allowed to not specify the context at all.

Regarding not passing `**kwargs`: I do not see a good reason for this.

**Comment by felixxm** [comment]
> There are now two overlapping ways to set the Content-Type header. I think it makes sense to take content_type over headers, but not sure if we should check this in __init__() and raise an exception, document it, or perhaps neither is necessary.

I think we should raise a `ValueError` that they're mutually exclusive, and maybe deprecate `content_type`.

**Review by felixxm — COMMENTED** [review]
@knyghty Thanks :+1:

I think we should also update ["Telling the browser to treat the response as a file attachment"](https://docs.djangoproject.com/en/3.1/ref/request-response/#telling-the-browser-to-treat-the-response-as-a-file-attachment) section and pass `Content-Type` and `Content-Disposition` in `headers`.

**Comment by carltongibson** [comment]
I think we should leave `content_type`. Just as before, it's a handy shortcut for (probably) the most common header. Plus **much code** will be using it. 

>I think we should raise a ValueError that they're mutually exclusive...

Yes, 👍. If `content_type is not None` we should `raise` if `Content-Type` (case-insensitive) in `headers`.

**Comment by knyghty** [comment]
I'd love to deprecate it, but I agree with your Carlton, I think it would cause too many unnecessary complaints for too little gain.

I will add the check though, and update those docs.

@felixxm, @carltongibson, what do you both think about adding `**kwargs` to `TemplateResponse` and `SimpleTemplateResponse`? It seems like a good idea to me.

**Comment by carltongibson** [comment]
> @felixxm, @carltongibson, what do you both think about adding **kwargs to TemplateResponse and SimpleTemplateResponse? It seems like a good idea to me.

I see the case for adding `headers` here. I'd think changing to `**kwargs` would be out of scope for this PR (and something we'd want to think about... — it's not always the case that the broader API is better...)

**Comment by knyghty** [comment]
Thanks both, addressed those and fixed an issue where content type wouldn't be set from the headers dict. Also moved the code closer together to make it more understandable.

File: tests/template_tests/test_response.py
Comment (felixxm) [inline tests/template_tests/test_response.py:225]: Do we need to call `render()`? It looks unnecessary.

File: tests/template_tests/test_response.py
Comment (felixxm) [inline tests/template_tests/test_response.py:338]: Do we need to call `render()`? It looks unnecessary.

File: tests/template_tests/test_response.py
Comment (felixxm) [inline tests/template_tests/test_response.py:339]:
```suggestion
        self.assertEqual(response.headers['x-foo'], 'foo')
```

File: tests/template_tests/test_response.py
Comment (felixxm) [inline tests/template_tests/test_response.py:226]:
```suggestion
        self.assertEqual(response.headers['x-foo'], 'foo')
```

File: django/template/response.py
Comment (felixxm) [inline django/template/response.py:14]: [Docs](https://docs.djangoproject.com/en/3.1/ref/template-response/#django.template.response.SimpleTemplateResponse.__init__) changes are missing.

File: django/template/response.py
Comment (felixxm) [inline django/template/response.py:142]: [Docs](https://docs.djangoproject.com/en/3.1/ref/template-response/#django.template.response.TemplateResponse.__init__) changes are missing.

File: docs/releases/3.2.txt
Comment (felixxm) [inline docs/releases/3.2.txt:317]:
```
* The new ``headers`` parameter of ``:class:`~django.http.HttpResponse``,
  :class:`~django.template.response.SimpleTemplateResponse`, and
  :class:`~django.template.response.TemplateResponse` allows ...
```

File: tests/httpwrappers/tests.py
Comment (felixxm) [inline tests/httpwrappers/tests.py:486]:
```suggestion
        msg = ...
        with self.assertRaisesMessage(ValueError, msg):
```

**Review by felixxm — COMMENTED** [review]
@knyghty Thanks for updates :+1:

**Comment by knyghty** [comment]
Thanks @felixxm, I've resolved those and also added a test for the case I found yesterday.

File: tests/template_tests/test_response.py
Comment (felixxm) [inline tests/template_tests/test_response.py:225]: Please use `assertEqual()`  instead of `assert ... == ...`.

Comment (knyghty) [reply tests/template_tests/test_response.py]: Gah, sorry, force of habit, I'm too used to pytest these days.

File: tests/httpwrappers/tests.py
Comment (felixxm) [inline tests/httpwrappers/tests.py]: This test works without the patch, I will move it to a separate commit.

**Review by felixxm — COMMENTED** [review]
@knyghty Thanks for updates :+1: 

I added missing `versionchanged` annotations, added the ``headers`` parameter to ``HttpResponse.__init__()`` signature in docs, moved an extra test to a separate commit, and 
pushed minor edits.

**Review by carltongibson — APPROVED** [review]
OK, looks good. Thanks @knyghty  this'll be nice.

**Comment by felixxm** [comment]
Merged in 2e7cc95499f758a1c4aa036cbf1dcddf82a89ea2 and dcb69043d0de45bb55998fc418d93c28bc7689ae.

**Pull request closed without merging**
